float32 x
float32 y
float32 z
float32 normal_x
float32 normal_y
float32 normal_z